### PR TITLE
[IE CLDNN] Build time optimization for OCL kernels

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
- Build OCL kernel batches in parallel 
- Two implementations for threading : tbb & threadpool

### CODE REVIEW

**Comment by yeonbok** [comment]
@ababushk  Hi Andrey, this patch is adding tbb thread on cldnn code, but seems that cldnn unittest job is not loading tbb library. Could you please add tbb path to LD_LIBRARY_PATH for the environment of that job? 
https://openvino-ci.intel.com/job/private-ci/job/ie/job/ie-tests-linux-ubuntu18-cldnn_unit/28131/console

**Comment by yeonbok** [comment]
@vladimir-paramuzov Having two versions of threads : thread pool + tbb, we have a concern of management. How do you think remove threadpool path and go only with tbb?

**Comment by ababushk** [comment]
@yeonbok CI environment fix was merged to master, please make some new commits to your PR for changes to be propagated

**Comment by yeonbok** [comment]
> @yeonbok CI environment fix was merged to master, please make some new commits to your PR for changes to be propagated

Thanks! : )

File: inference-engine/src/cldnn_engine/cldnn_config.h [hunk 1]
```diff
@@ -56,6 +57,7 @@ struct Config {
     std::string sources_dumps_dir;
     std::string device_id;
     std::string kernels_cache_dir;
+    int n_threads;
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/src/cldnn_engine/cldnn_config.h:60]: we intended to make it to set  max # threads if user puts -1. 
but... I think 0 is okay too..

[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply inference-engine/src/cldnn_engine/cldnn_config.h:60]: I think it always has valid values here, hasn't it? If user set -1, then you convert it to max_threads in cldnn_config.cpp, and assign correct value to `n_threads` field.

File: inference-engine/thirdparty/clDNN/CMakeLists.txt [hunk 2]
```diff
@@ -69,6 +69,14 @@ set_property(GLOBAL PROPERTY USE_FOLDERS ON)
 # ======================================================================================================
 # ====================================== HELPER CONSTANT VARIABLES =====================================
 # ======================================================================================================
+# ======================================================================================================
+if("${THREADING}" MATCHES "SEQ")
+    add_definitions(-DCLDNN_OCL_BUILD_THREADING=CLDNN_OCL_BUILD_THREADING_SEQ)
+elseif("${CLDNN_THREADING}" MATCHES "TBB")
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/CMakeLists.txt:75]: it was mistake... thanks!

File: inference-engine/include/cldnn/cldnn_config.hpp [hunk 3]
```diff
@@ -72,6 +72,11 @@ DECLARE_CLDNN_CONFIG_KEY(ENABLE_FP16_FOR_QUANTIZED_MODELS);
 */
 DECLARE_CLDNN_CONFIG_KEY(NV12_TWO_INPUTS);
 
+/**
+* @brief This key sets the number of threads used to build program.
+* 2 is a default value.
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/include/cldnn/cldnn_config.hpp]: Applied, thanks!

File: inference-engine/thirdparty/clDNN/CMakeLists.txt [hunk 4]
```diff
@@ -69,6 +69,14 @@ set_property(GLOBAL PROPERTY USE_FOLDERS ON)
...
+    add_definitions(-DCLDNN_OCL_BUILD_THREADING=CLDNN_OCL_BUILD_THREADING_TBB)
+else()
+    add_definitions(-DCLDNN_OCL_BUILD_THREADING=CLDNN_OCL_BUILD_THREADING_THREADPOOL)
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/CMakeLists.txt]: Applied thanks,

File: inference-engine/src/cldnn_engine/cldnn_config.cpp [hunk 5]
```diff
@@ -221,6 +222,20 @@ void Config::UpdateFromMap(const std::map<std::string, std::string>& configMap)
             } else {
                 THROW_IE_EXCEPTION_WITH_STATUS(NotFound) << "Unsupported KEY_CLDNN_ENABLE_FP16_FOR_QUANTIZED_MODELS flag value: " << val;
             }
+        } else if (key.compare(CLDNNConfigParams::KEY_CLDNN_MAX_NUM_THREADS) == 0) {
+            int max_threads = (std::thread::hardware_concurrency() == 0) ? 1 : std::thread::hardware_concurrency();
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/src/cldnn_engine/cldnn_config.cpp]: Applied!

File: inference-engine/src/cldnn_engine/cldnn_config.h [hunk 6]
```diff
@@ -31,7 +31,8 @@ struct Config {
                graph_dumps_dir(""),
                sources_dumps_dir(""),
                device_id(""),
-               kernels_cache_dir("") {
+               kernels_cache_dir(""),
+               n_threads((std::thread::hardware_concurrency() == 0) ? 1 : std::thread::hardware_concurrency()) {
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/src/cldnn_engine/cldnn_config.h]: Applied

File: inference-engine/src/cldnn_engine/cldnn_config.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/src/cldnn_engine/cldnn_config.h]: Agree. Applied. Thanks.

File: inference-engine/thirdparty/clDNN/api/engine.hpp [hunk 7]
```diff
@@ -61,6 +62,7 @@ struct engine_configuration {
                                               ///< (switched off for older drivers then NEO).
     uint16_t n_streams;                       ///< Number of queues executed in parallel
     const std::string kernels_cache_path;     ///< Path to compiled kernels cache
+    uint16_t n_threads; ///< Number of threads used to build program
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/api/engine.hpp:65]: I suppose it makes sense to align this comment as others.

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.h [hunk 8]
```diff
@@ -26,14 +39,76 @@ namespace cldnn {
...
+    std::queue<std::function<void()>> _tasks;
+    std::condition_variable _cv;
+    std::mutex _q_m;
+    bool _stop_pool;
+
+    void worker_thread() {
+        while (true) {
+            std::unique_lock<std::mutex> lock(this->_q_m);
+            _cv.wait(lock, [this]() { return (!this->_tasks.empty()) || (_stop_pool); });
+            if ( (_stop_pool) && (this->_tasks.empty())) return;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.h:95]: Please remove extra space before break.

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp [hunk 9]
```diff
@@ -313,149 +337,160 @@ static std::vector<unsigned char> getProgramBinaries(cl::Program program) {
...
+    return res->second;
 }
 
 void kernels_cache::build_all() {
     OV_ITT_SCOPED_TASK(itt::domains::CLDNN, "KernelsCache::BuildAll");
     if (!_pending_compilation)
         return;
+    std::vector<batch_program> batches;
+    {
+        std::lock_guard<std::mutex> lock(_context.get_cache_mutex());
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:449]: Yes, we have. But some test cases are disabled on GPU 😄 https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/tests/functional/plugin/gpu/shared_tests_instances/skip_tests_config.cpp#L14-L15

File: inference-engine/thirdparty/clDNN/api/engine.hpp (hunk 7 above)
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/api/engine.hpp]: Applied

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.h (hunk 8 above)
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.h]: Applied

File: inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:450]: Why are they skipped? Then is there still running tests for threads?

[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:450]: See Jira 39014

[the remark this answers is not in the document]
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/src/gpu/kernels_cache.cpp:450]: @vladimir-paramuzov  I tried loading two models at a time (googlenet / resnet) in my local, and it worked well.

**Review by sshlyapn — APPROVED** [review]

**Pull request merged**